tests/testsuite/workspaces.rs

```diff
@@ -1628,6 +1628,94 @@ fn exclude_but_also_depend() {
     assert!(p.root().join("foo/bar/target").is_dir());
 }
 
+#[cargo_test]
+fn excluded_default_members_still_must_be_members() {
+    let p = project()
+        .file(
+            "Cargo.toml",
+            r#"
+            [workspace]
+            members = ["foo"]
+            default-members = ["foo", "bar"]
+            exclude = ["bar"]
+        "#,
+        )
+        .file("foo/Cargo.toml", &basic_manifest("foo", "0.1.0"))
+        .file("foo/src/lib.rs", "")
+        .file("bar/something.txt", "");
+    let p = p.build();
+    p.cargo("build")
+        .with_status(101)
+        .with_stderr(
+            "\
+error: package `[..]bar` is listed in workspace’s default-members \
+but is not a member.
+",
+        )
+        .run();
+}
+
+#[cargo_test]
+fn excluded_default_members_crate_glob() {
+    let p = project()
+        .file(
+            "Cargo.toml",
+            r#"
+            [workspace]
+            members = ["foo", "bar/*"]
+            default-members = ["bar/*"]
+            exclude = ["bar/quux"]
+        "#,
+        )
+        .file("foo/Cargo.toml", &basic_manifest("foo", "0.1.0"))
+        .file("foo/src/main.rs", "fn main() {}")
+        .file("bar/baz/Cargo.toml", &basic_manifest("baz", "0.1.0"))
+        .file("bar/baz/src/main.rs", "fn main() {}")
+        .file("bar/quux/Cargo.toml", &basic_manifest("quux", "0.1.0"))
+        .file("bar/quux/src/lib.rs", "");
+
+    let p = p.build();
+    p.cargo("build").run();
+
+    assert!(p.root().join("target").is_dir());
+    assert!(!p.bin("foo").is_file());
+    assert!(p.bin("baz").is_file());
+
```

Comment:
Maybe also add a check `!p.bin("quux").exists()` here and the `--workspace` case?

Reply:
Done.